enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,188 @@
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+### Drawbacks
+* External certificate wont be automaticly renewed, so it requires manual Certificate rotation. 
+  similiar approach is taken by OpenShift.
+
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+- How to handle kubeconfig generation for NamedCerts?
+
+### Test Plan
+add e2e test that will:
+- generate and sign certificates with custom ca.
+- change the default Microshift configuration to use the newly generated certs.
+- make sure system is functional using  generated external kubeconfig.
+
+### Graduation Criteria
+#### Dev Preview -> Tech Preview
+- Gather feedback from early adopters on possible issues.
+
+#### Tech Preview -> GA
+- Extensive testing with various certificates variations.
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+the provided certs value will be validated before is it passed to the api-server flag
+
+this check will prevent Microshift service from starting:
+1. certificates files exists in the disk and readable by microshift process.
+1. certificates shouldnt override the internal local certificates which are managed internally.
+
+this check display warning message at the log and service will be started:
+1. certificates is expired.
+
+
+#### Support Procedures
+configured certs value are prepended to the  `tls-sni-cert-key`  TLS handshake command line flag  which is passed to the kube-apiserver:
+```
+> journalctl -u microshift -b0 | grep tls-sni-cert-key
+
+Jan 24 14:53:00 localhost.localdomain microshift[45313]: kube-apiserver I0124 14:53:00.649099   45313 flags.go:64] FLAG: --tls-sni-cert-key="[/home/eslutsky/dev/certs/server.crt,/home/eslutsky/dev/certs/server.key;/var/lib/microshift/certs/kube-apiserver-external-signer/kube-external-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-external-signer/kube-external-serving/server.key;/var/lib/microshift/certs/kube-apiserver-localhost-signer/kube-apiserver-localhost-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-localhost-signer/kube-apiserver-localhost-serving/server.key;/var/lib/microshift/certs/kube-apiserver-service-network-signer/kube-apiserver-service-network-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-service-network-signer/kube-apiserver-service-network-serving/server.key
+```
+
+Connecting to API server using external SNI yields external certificate.
+   ```shell
+   $ openssl s_client -connect <SNI_ADDRESS>:6443 -showcerts | openssl x509 -text -noout -in - | grep -C 1 "Alternative\|CN"
+   ```
+
+## Implementation History
+N/A
+
+## Alternatives
+- Use cert-manager for managing Microshift external custom certs
+  - Benifits
+    1. will allow Microshift Administrator create automatic Certificate rotation strategy using supported protocols  (ie: ACME Webhook)
```

Comment:
Will allow MicroShift administrators to rotate certificates automatically (e.g. via ACME)